packages/zwave-js/src/lib/controller/Controller.ts

```diff
@@ -1870,23 +1873,24 @@ ${associatedNodes.join(", ")}`,
 				ZWaveErrorCodes.ReplaceFailedNode_Failed,
 			);
 		} else {
-			switch (result.removeStatus) {
+			switch (result.replaceStatus) {
 				case ReplaceFailedNodeStatus.NodeOK:
 					throw new ZWaveError(
-						`The node could not be removed because it has responded`,
+						`The node could not be replaced because it has responded`,
 						ZWaveErrorCodes.ReplaceFailedNode_NodeOK,
 					);
 				case ReplaceFailedNodeStatus.FailedNodeReplaceFailed:
 					throw new ZWaveError(
 						`The removal process could not be completed`,
 						ZWaveErrorCodes.ReplaceFailedNode_Failed,
 					);
+				case ReplaceFailedNodeStatus.FailedNodeReplace:
+					// failed node is now ready to be replaced and controller is ready to add a new
+					// node with the nodeID of the failed node
+					return this.beginInclusion(includeNonSecure);
 				default:
-					// If everything went well, the status is RemoveFailedNodeStatus.NodeRemoved
-
-					// TODO: wtf
-
-					return;
+					// If everything went well, the status is ReplaceFailedNodeStatus.FailedNodeReplaceDone
+					return true;
```

Comment:
We need to figure out when which message comes. If responses may be received multiple times (like during the inclusion process) we might need to create a deferred promise to be returned here.

I just remembered that I had a log of that from OZW - it might give you some pointers :)
==> https://github.com/zwave-js/node-zwave-js/blob/master/specs/OZW_Log__replace_node.txt

Reply:
Nevermind, that file doesn't actually include anything useful - I should have taken a look at it when I made it back then 😅

Reply:
> We need to figure out when which message comes. If responses may be received multiple times (like during the inclusion process) we might need to create a deferred promise to be returned here. 

My zwave understanding at the low level is very low but I think that this could return twice: firstly returns when `beginInclusion` promise is fulfilled and then when we receive a response `ReplaceFailedNodeStatus.FailedNodeReplaceDone`

Reply:
>  we might need to create a deferred promise to be returned here

You mean the beginInclusionPromise or a new replaceFailed promise?

Reply:
From what I have understanded It could work like this:

---> ReplaceFailedNodeRequest
<--- ReplaceFailedNodeStatus.FailedNodeReplace
---> AddNodeToNetworkRequest
<--- ReplaceFailedNodeStatus.FailedNodeReplaceDone

Reply:
I think it might work like this:
```
---> ReplaceFailedNodeRequest (start the process)
<--- ReplaceFailedNodeStatus.FailedNodeReplace OR an error (you've handled the error case)
<--- AddNodeToNetworkRequest (MULTIPLE!)
<--- ReplaceFailedNodeStatus.FailedNodeReplaceDone OR Failed
```

The promise returned by the controller method should be rejected in case of an error (either before or during the replacing) or it should be resolved after the node was added (status ReplaceDone)

To be able to handle multiple incoming messages, please take a look at how I've handled the incoming AddNodeToNetworkRequests in https://github.com/zwave-js/node-zwave-js/blob/4dffb8de6796723604da87e26cc8f8bf8ab0b434/packages/zwave-js/src/lib/controller/Controller.ts#L147-L159

Reply:
Actually this would be perfect for [state machines](https://github.com/zwave-js/node-zwave-js/issues/884), but this is a whole other beast.